Python/parking_lot.c

```diff
@@ -0,0 +1,470 @@
+{
+    struct llist_node *root = &bucket->root;
+    if (!root->next) {
+        // bucket was not yet initialized
+        return;
+    }
+
+    // remove and append all matching waiters to dst
+    struct llist_node *node;
+    llist_for_each_safe(node, root) {
+        struct wait_entry *wait = llist_data(node, struct wait_entry, node);
+        if (wait->addr == (uintptr_t)address) {
+            llist_remove(node);
+            llist_insert_tail(dst, node);
+            --bucket->num_waiters;
+        }
+    }
+}
+
+// Checks that `*addr == *expected`
+static int
+validate_addr(const void *addr, const void *expected, size_t addr_size)
+{
+    switch (addr_size) {
+    case 1: return _Py_atomic_load_uint8(addr) == *(const uint8_t *)expected;
+    case 2: return _Py_atomic_load_uint16(addr) == *(const uint16_t *)expected;
+    case 4: return _Py_atomic_load_uint32(addr) == *(const uint32_t *)expected;
+    case 8: return _Py_atomic_load_uint64(addr) == *(const uint64_t *)expected;
+    default: Py_UNREACHABLE();
+    }
+}
+
+int
+_PyParkingLot_Park(const void *addr, const void *expected, size_t size,
+                   _PyTime_t timeout_ns, void *arg, int detach)
+{
+    struct wait_entry wait;
+    wait.unpark.arg = arg;
+    wait.addr = (uintptr_t)addr;
+
+    Bucket *bucket = &buckets[((uintptr_t)addr) % NUM_BUCKETS];
+
+    _PyRawMutex_Lock(&bucket->mutex);
+    if (!validate_addr(addr, expected, size)) {
+        _PyRawMutex_Unlock(&bucket->mutex);
+        return Py_PARK_AGAIN;
+    }
+    wait.sema = _PySemaphore_Alloc();
+    enqueue(bucket, addr, &wait);
+    _PyRawMutex_Unlock(&bucket->mutex);
+
+    int res = _PySemaphore_Wait(wait.sema, timeout_ns, detach);
+    if (res == Py_PARK_OK) {
+        goto done;
+    }
+
+    // timeout or interrupt
+    _PyRawMutex_Lock(&bucket->mutex);
+    if (wait.node.next == NULL) {
+        _PyRawMutex_Unlock(&bucket->mutex);
+        // We've been removed the waiter queue. Wait until we process the
+        // wakeup signal.
+        do {
+            res = _PySemaphore_Wait(wait.sema, -1, detach);
+        } while (res != Py_PARK_OK);
+        goto done;
+    }
+    else {
+        llist_remove(&wait.node);
+        --bucket->num_waiters;
+    }
+    _PyRawMutex_Unlock(&bucket->mutex);
+
+done:
+    _PySemaphore_Free(wait.sema);
+    return res;
+
+}
+
+struct _PyUnpark *
+_PyParkingLot_BeginUnpark(const void *addr)
+{
+    Bucket *bucket = &buckets[((uintptr_t)addr) % NUM_BUCKETS];
+    _PyRawMutex_Lock(&bucket->mutex);
+
+    struct wait_entry *waiter = dequeue(bucket, addr);
+    if (!waiter) {
+        return NULL;
+    }
+
+    waiter->unpark.has_more_waiters = (bucket->num_waiters > 0);
+    return &waiter->unpark;
+}
+
+#define container_of(ptr, type, member) \
+    ((type *)((char *)(ptr) - offsetof(type, member)))
+
+void
+_PyParkingLot_FinishUnpark(const void *addr, struct _PyUnpark *unpark)
+{
+    Bucket *bucket = &buckets[((uintptr_t)addr) % NUM_BUCKETS];
```

Comment:
If this is always paired with `_PyParkingLot_BeginUnpark()`, we shouldn't need to look up the bucket again:

```C
Bucket *bucket;
struct _PyUnpark *unpark = _PyParkingLot_BeginUnpark(addr, &bucket);
...
_PyParkingLot_FinishUnpark(bucket, unpark);
```

Reply:
Yeah, that the function pointer looks great.